### PR TITLE
Empty dashboard wired into routing

### PR DESCRIPTION
**Pull request by zacharyhsia**
Add dashboard page to routing and set a flag in environment files.

### CODE REVIEW

File: ui/src/app/app.module.ts [hunk 1]
```diff
@@ -11,6 +11,8 @@ import {JobDetailsModule} from './job-details/job-details.module';
 import {JobListModule} from './job-list/job-list.module';
 import {SignInModule} from './sign-in/sign-in.module';
 import {ProjectsModule} from './projects/projects.module';
+import { DashboardComponent } from './dashboard/dashboard.component';
```
Comment (calbach) [inline ui/src/app/app.module.ts]: nit: drop the surrounding spaces to match the rest of the style

File: ui/src/app/dashboard/dashboard.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+describe('DashboardComponent', () => {
+  let component: DashboardComponent;
+  let fixture: ComponentFixture<DashboardComponent>;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [ DashboardComponent ]
+    })
+    .compileComponents();
```
Comment (calbach) [inline ui/src/app/dashboard/dashboard.component.spec.ts]: Generally should indent for continuations like this. Alternatively in this case I would probably put it on the previous line

**Comment by zacharyhsia** [comment]
Please review the PR. Thanks!

File: ui/src/app/projects/projects.component.ts [hunk 3]
```diff
@@ -94,6 +95,10 @@ export class ProjectsComponent implements OnInit {
 
   navigateJobs() {
     let extras = {queryParams: {q: URLSearchParamsUtils.encodeURLSearchParams({extensions: {projectId: this.projectsControl.value}})}};
-    this.router.navigate(['jobs'], extras);
+    if (environment.dashboardEnabled) {
+      this.router.navigate(['dashboard'], extras);
```
Comment (calbach) [inline ui/src/app/projects/projects.component.ts]: Have you thought about the URL scheme? I'm thinking we may just want `projectId=<project>`

File: ui/src/app/app-routing.module.ts [hunk 4]
```diff
@@ -34,6 +35,11 @@ const routes: Routes = [
     component: ProjectsComponent,
     canActivate: [CapabilitiesActivator]
   },
+  {
+    path: 'dashboard',
+    component: DashboardComponent,
+    canActivate: [CapabilitiesActivator]
```
Comment (calbach) [inline ui/src/app/app-routing.module.ts]: One thing we'll need to handle in the CapabilitiesActivator: the dsub dashboard page will require a project (probably via query parameter). If it's not there, we'd want to redirect to the projects page (similar to how we're redirecting from "jobs" -> "projects" if there is no projectId in the query).

Feel free to handle this in this PR, or else just put a TODO in the CapabilitiesActivator

File: ui/src/app/projects/projects.component.ts (hunk 3 above)
Comment (zacharyhsia) [reply ui/src/app/projects/projects.component.ts]: I thought we can parse that 'extra' thing when handling the navigation. It is not that big.

Comment (calbach) [reply ui/src/app/projects/projects.component.ts]: I think the only filter we would be supporting here is projectId (and possibly a query param for the time range), so I wouldn't support the more generic "?q=<encoded_search_query>" that we have on the job list page.

File: ui/src/app/shared/utils/url-search-params.utils.ts [hunk 5]
```diff
@@ -6,7 +6,7 @@ import {JobStatus} from "../model/JobStatus";
 import {FieldDataType, queryDataTypes, queryExtensionsDataTypes} from "../common";
 
 /** Utilities for working with URLSearchParams*/
-export class URLSearchParamsUtils {
+  export class URLSearchParamsUtils {
```
Comment (calbach) [inline ui/src/app/shared/utils/url-search-params.utils.ts]: Revert

File: ui/src/app/projects/projects.component.ts [hunk 6]
```diff
@@ -93,7 +94,13 @@ export class ProjectsComponent implements OnInit {
   }
 
   navigateJobs() {
-    let extras = {queryParams: {q: URLSearchParamsUtils.encodeURLSearchParams({extensions: {projectId: this.projectsControl.value}})}};
-    this.router.navigate(['jobs'], extras);
+    let jobsExtras = {queryParams: {q: URLSearchParamsUtils.encodeURLSearchParams({extensions: {projectId: this.projectsControl.value}})}};
+    let dashboardExtras = {queryParams: {projectId: this.projectsControl.value}};
```
Comment (calbach) [inline ui/src/app/projects/projects.component.ts]: nit: I would prefer inlining these below or else defining the variables right before they are used.

Comment (calbach) [inline ui/src/app/projects/projects.component.ts]: opt: existing issue, but these should really be const's

**Comment by zacharyhsia** [comment]
I added another extra with only projectId as param. 
How can I add multiple param? 
Here is what I found on angular doc,
```javascript
interface NavigationExtras {
  relativeTo?: ActivatedRoute | null
  queryParams?: Params | null
  fragment?: string
  preserveQueryParams?: boolean
  queryParamsHandling?: QueryParamsHandling | null
  preserveFragment?: boolean
  skipLocationChange?: boolean
  replaceUrl?: boolean
}
```

```javascript
type Params = {
    [key: string]: any;
};
```

I am confused by `[key: string]: any;`, I thought the following code would fit in but it dose not.
```javascript
let dashboardExtras = {queryParams: [{projectId: this.projectsControl.value}, {anotherField: 'value'}]};
```

How to interpret `[key: string]: any;`

**Comment by rexwangcc** [comment]
@calbach Thanks for the information here!

**Pull request merged**